### ISSUE TITLE
Support setting tedge configuration via environment variables

### ISSUE THREAD

**Issue by reubenmiller** [thin-edge/thin-edge.io#1783]
**Is your feature improvement request related to a problem? Please describe.**

It is difficult to configure thin-edge.io when using within a container (multi-process or single process container setups) as the configuration is file based, which normally requires users to create a custom init script, which writes the values to file before starting the service.

**Describe the solution you'd like**

**Update 2023-03-23: Updated spec due to further improvement #1837**

Support controlling values from the `tedge.toml` using environment variables.

The environment variable name translation should:
* Convert setting name to uppercase
* Replacing all dots (`.`) with underscore (`_`)
* Use a common prefix `TEDGE_` to namespace each variable

For example the mapping between setting key and environment variable COULD look like this:

|Setting|Environment variable|
|-------|-------------------------------|
|`c8y.url`|`TEDGE_C8Y_URL`|
|`device.key.path`|`TEDGE_DEVICE_KEY_PATH`|
|`device.cert.path`|`TEDGE_DEVICE_CERT_PATH`|

**Configuration layering precedence**

Since the settings can be set via multiple sources, the following details the precedence of settings, and the the first non-empty value should be used.

1. Read value from Environment variables (treat an empty value as `not-set`)
2. Read value from `tedge.toml`
3. Read from code default

**User facing example**

Once this is supported it would make it possible building a container image which can set all of the required values when starting the container, and the thin-edge component binary can be called directly (without requiring an intermediate step):

```sh
docker run \
    --env TEDGE_C8Y_URL=myurl.cumulocity.com \
    --env TEDGE_DEVICE_KEY_PATH=/cert-volume/tedge.key \
    --env TEDGE_DEVICE_CERT_PATH=/cert-shared-volume/tedge.pem \
    myimage \
    tedge-agent
```

**Describe alternatives you've considered**

**Additional context**

This pattern is very common and is supported by various RUST libraries such as:

* https://github.com/mehcode/config-rs
* https://github.com/SergioBenitez/Figment

**Comment by jarhodes314** [comment]
Using `_` as the separator is easy for simple cases such as `C8Y_URL`, but not fore `DEVICE_KEY_PATH`, which would translate to `device.key.path`, whereas we want `device.key_path`. Both of the libraries resolve this ambiguity (by default) by using `.` instead of `_` as the separator. In general, they're following a two step process: convert the variables to some sensible default data structure, such as `{c8y: {url: "myurl.cumulocity.com"}}` and then attempt to convert this value into our configuration struct. If we want to use `_` as the separator, but also have `_` in the configuration keys, we'd basically just have to either:

(a): keep guessing until we stumble across a valid combination of `.` and `_`
(b): hardcode each of these variables in

If we were to use `C8Y.URL` and `DEVICE.KEY_PATH` as the variable names, that would make everything a lot easier (as we can just use the library rather than coming up with the logic ourselves).

Another thing to bear in mind is whether we want to prefix the variables (e.g. `TEDGE_C8Y.URL` or `TEDGE.C8Y.URL`). It probably doesn't matter in the context of a container, but it might make things easier if there's a use case for this configuration elsewhere, where we don't necessarily want to accidentally collide with an existing variable (although the use of `.` makes that unlikely).

I also find the idea of a config file overriding environment variables to be surprising, I'd expect it to be the exact opposite as I don't have to persist environment variable changes, but I do have to persist config changes.

**Comment by reubenmiller** [comment]
We can use a double underscore separator or just forbid people to use underscores in a property name

**Comment by reubenmiller** [comment]
> (a): keep guessing until we stumble across a valid combination of . and _

Environment variables should not use `.` in them. Whilst technically it might work in some occassions, it is considered bad practice or not against linux spec. [Reference](https://www.baeldung.com/linux/allowed-characters-variable-names)

**Comment by Bravo555** [comment]
When working on #1773, I was also surprised to find that some keys like `mqtt.external.port` are saved into `/etc/tedge/tedge.toml` like this:

```toml
[mqtt]
external_port = 1234
```

instead of something like:

```toml
[mqtt.external]
port = 1234
```

or

```
[mqtt]
external.port = 1234
```

which I think is a little misleading. The first `.` denoted a table, why wouldn't the second? Not to mention [TOML supports using dots inside keys (dotted keys)](https://toml.io/en/v1.0.0#keys). So we could theoretically replace all the `_` in keys with `.`, which would make all keys trivially translatable to env variables, but we would need to add an alias for every field in DTO object that contains underscores.

It seems to me like we can't decide on an config structure and want to have it both ways. The major goal of TOML is to map unambiguously to a deep hash table, so we created structs to deserialize to, but then we made [an API for accessing this config as a flat typemap](https://github.com/thin-edge/thin-edge.io/blob/main/crates/common/tedge_config/src/tedge_config_cli/settings.rs), where you have to add trivial getters/setters for every field, and manually define a key that is used in CLI, but has no relation to how it will actually be saved in a config file. Because of that, adding new config settings is a pain, and I believe it could be made a bit simpler.

> This pattern is very common and is supported by various RUST libraries such as:
> 
> - https://github.com/mehcode/config-rs
> - https://github.com/SergioBenitez/Figment

I'd be for looking into if we could make use of some dedicated config crate, so we could eliminate some of the boilerplate that we have wrt. configuration options. But this needs to be considered in the context of our architecture and our past design decisions, keeping in mind who can write which configuration options, who can read which options, etc. and if these common solutions will fit our needs, so this should probably be a subject of another ticket.

**Comment by didier-wenzek** [comment]
> I believe it could be made a bit simpler

Unfortunately true.

This crate has been implemented at a time when the requirements were full of constraints (as exposing this file to the user *only* with the  `tedge config` command - hiding the file and even its location). We added on purpose, but then abused, this ability to change our mind on a name exposed to the user without breaking a former configuration file. Meantime most of these constraints have been removed, but the abstractions are still here adding a lot of burden to add a config parameter or override those with environment variables as proposed here.

> I'd be for looking into if we could make use of some dedicated config crate, so we could eliminate some of the boilerplate that we have wrt. configuration options. But this needs to be considered in the context of our architecture and our past design decisions, keeping in mind who can write which configuration options, who can read which options, etc. and if these common solutions will fit our needs, so this should probably be a subject of another ticket.

I would support such an effort. We need to stay calm though; and take some time to assess what we really need and what has to be done to convert old configuration files.

**Comment by didier-wenzek** [comment]
> I'd be for looking into if we could make use of some dedicated config crate, so we could eliminate some of the boilerplate that we have wrt. configuration options. But this needs to be considered in the context of our architecture and our past design decisions, keeping in mind who can write which configuration options, who can read which options, etc. and if these common solutions will fit our needs, so this should probably be a subject of another ticket.

Beyond the boilerplate to add the settings for MQTT clients address and port, what bothers me in this PR  
https://github.com/thin-edge/thin-edge.io/pull/1789, is the number of time we repeat ourselves to extract the MQTT connection settings from tedge.conf! This should be done in a single place!

**Comment by jarhodes314** [comment]
I agree that this needs considering quite carefully before we attempt to fix the more structural issues. After all, it's near enough entirely accidental complexity in this case. The library I used for the environment variable handling, `figment`, is only really designed for reading the configuration files, and that's all I've tried to do with it here. At a minimum, we need to support derived/inferred "configuration" (such as the device id from the certificate) and writing to variables using `tedge config set`.

I still need to add documentation for the environment variables, but there is a lot of potential confusion as to the arbitrary relationship between config keys and dto keys. IMO it would make sense to preserve the structure of the underlying configuration, but when using `tedge config`, deprecate the use of e.g. `mqtt.external.port` in favour of `mqtt.external_port`.
